datafusion/physical-plan/src/repartition/on_demand_repartition.rs

```diff
@@ -0,0 +1,1589 @@
+/// partitions until the output partitions request the data.
+///
+/// When polling, the operator sends the output partition number to the one partition channel, then the prefetch buffer will distribute the data based on the order of the partition number.
+/// Each input steams has a prefetch buffer(channel) to distribute the data to the output partitions.
+///
+/// The following diagram illustrates the data flow of the OnDemandRepartitionExec operator with 3 output partitions for the input stream 1:
+/// ```text
+///         /\                     /\                     /\
+///         ││                     ││                     ││
+///         ││                     ││                     ││
+///         ││                     ││                     ││
+/// ┌───────┴┴────────┐    ┌───────┴┴────────┐    ┌───────┴┴────────┐
+/// │     Stream      │    │     Stream      │    │     Stream      │
+/// │       (1)       │    │       (2)       │    │       (3)       │
+/// └────────┬────────┘    └───────┬─────────┘    └────────┬────────┘
+///          │                     │                       │    / \
+///          │                     │                       │    | |
+///          │                     │                       │    | |
+///          └────────────────┐    │    ┌──────────────────┘    | |
+///                           │    │    │                       | |
+///                           ▼    ▼    ▼                       | |
+///                       ┌─────────────────┐                   | |
+///  Send the partition   │ partion channel │                   | |
+///  number when polling  │                 │                   | |
+///                       └────────┬────────┘                   | |
+///                                │                            | |
+///                                │                            | |
+///                                │  Get the partition number  | |
+///                                ▼  then send data            | |
+///                       ┌─────────────────┐                   | |
+///                       │ Prefetch Buffer │───────────────────┘ |
+///                       │       (1)       │─────────────────────┘
+///                       └─────────────────┘ Distribute data to the output partitions
+///
+/// ```
+type OnDemandDistributionSender = tokio::sync::mpsc::UnboundedSender<MaybeBatch>;
+type OnDemandDistributionReceiver = tokio::sync::mpsc::UnboundedReceiver<MaybeBatch>;
+
+type OnDemandInputPartitionsToCurrentPartitionSender = Vec<OnDemandDistributionSender>;
+type OnDemandInputPartitionsToCurrentPartitionReceiver =
+    Vec<OnDemandDistributionReceiver>;
+/// Inner state of [`OnDemandRepartitionExec`].
+#[derive(Debug)]
+struct OnDemandRepartitionExecState {
+    /// Channels for sending batches from input partitions to output partitions.
+    /// Key is the partition number.
+    channels: HashMap<
```

Comment:
Is a `HashMap` needed here? It could be a `Vec`?

Reply:
No, it cannot be `Vec`, because we need the hashmap to keep the partitions one-to-one with the receivers.

https://github.com/Weijun-H/arrow-datafusion/blob/ad720776ee39980b619d663acde59291e04b5fce/datafusion/physical-plan/src/repartition/on_demand_repartition.rs#L438-L452